tools/query-udevadm.py

```diff
@@ -0,0 +1,162 @@
+#!/usr/bin/env python3
+
+'''
+Print a json with information about usb/udev devices. The devices are selected using
+an input --pattern. This program will return a json describing the usb devices according 
+to the command `udevadm info <device>`. Information is collected for each udev device 
+matching the input pattern.
+
+ex. `python3 query-udevadm.py --pattern=/dev/video*`
+`python3 query-udevadm.py --pattern=/dev/disk/by-id/* --indent=2`
+
+The patterns are matched according to the python glob module. This only provides support 
+for simple unix style patterns.
+
+See also, for example, (or similar):
+`udevadm info /dev/sda | sed -n 's/E: //p'`
+
+The json format is:
+{
+    "devices":[
+            {
+                "path":<pattern match path>
+                "companionDevice": <known/expected devices in companion>
+                "udevInfo":{
+                    <udev-attr>:<udev-value>,
+                    ...
+                }
+            },
+            ... 
+        ]
+    }
+}
+'''
+
+# TODO should we generally move os to subprocess in companion?
+# TODO python2 support
+import sys
+import subprocess
+import glob
+import argparse
+import json
+
+PARSER = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawTextHelpFormatter)
+PARSER.add_argument('--pattern',
+                    action="store",
+                    type=str,
+                    default="/dev/serial/by-id/*",
+                    help="Path to search for usb devices."
+                    )
+PARSER.add_argument('--indent',
+                    action="store",
+                    type=int,
+                    default=None,
+                    help="Indent level for json output formatting."
+                    )
+ARGS = PARSER.parse_args()
+
+companionFamiliarDevices = {
+    "Pixhawk Autopilot":
+    {
+        "ID_MODEL" : "PX4_FMU_v2.x",
+        "ID_SERIAL" : "3D_Robotics_PX4_FMU_v2.x_0"
+    },
+    "Pixhawk Autopilot Bootloader":
+    {
+        "ID_MODEL" : 'PX4_BL_FMU_v2.x',
+        "ID_SERIAL" : "3D_Robotics_PX4_BL_FMU_v2.x_0"
+    },
+    "Blue Robotics HD Low Light USB Camera":
+    {
+        "ID_VENDOR_ID" : "05a3",
+        "ID_MODEL_ID" : "9422"
+    },
+    "Raspberry Pi Camera Module":
+    {
+        "ID_V4L_PRODUCT" : "mmal service 16.1",
+        "COLORD_KIND" : "camera"
+    },
+    "Logitech c920 Camera":
+    {
+        "ID_MODEL_ID" : "082d",
+        "ID_VENDOR_ID" : "046d"
+    }
+}
```

Comment:
I think that would be better to put the devices in a list
and for each list to have a name, path, and others necessary fields. This will make it easier to interact, getting each json without necessarily getting the keys.

Reply:
I know what you are suggesting, but why is this better? It is necessary to get/use the keys, that's the reason for the existence of this list. I think the change would result in similar code lines and length.

Reply:
The keys are unique identifiers, if we move them to a 'name' in the json, then there may be duplicates, which does not make sense.

Reply:
Can you please give an example of what you are suggesting?

I don't understand this:
> Also, making each component with different keys inside makes harder to maintain, I think that it makes more sense to thing how to make it more generic.

The identifiers that I put are some of the minimum information to identify the different devices uniquely, not all devices use usb vid/pid etc, and some (like the pixhawk) use the same vid/pid, so need to be identified with some other key.

Reply:
> for name in names:
   names[name]...

This is just iterating over a dictionary. There are better ways: https://stackoverflow.com/a/3294899

but it's a normal thing.